streamer/pulp/streamer/cache.py

```diff
@@ -0,0 +1,167 @@
+        _lock (RLock): The object mutex.
+        _inventory (dict): The inventory of cached objects.
+            Each value is an Item.
+    """
+
+    def __init__(self, eviction_threshold=None):
+        """
+        Args:
+            eviction_threshold (timedelta): How long an unrequested item will be cached.
+
+        """
+        self.eviction_threshold = eviction_threshold or timedelta(hours=4)
+        self._lock = RLock()
+        self._inventory = {}
+
+    def add(self, key, object):
+        """
+        Add an object to the cache.
+
+        Args:
+            key (hashable): The caching key.
+            object (object): An object to be cached.
+        """
+        with self._lock:
+            self._inventory[key] = Item(object)
+
+    def purge(self, key):
+        """
+        Purge (delete) objects cached using the specified key.
+
+        Args:
+            key (hashable): The caching key.
+        """
+        with self._lock:
+            return self._inventory.pop(key)
+
+    def get(self, key):
+        """
+        Get a cached object by key.
+
+        Args:
+            key (hashable): The caching key.
+
+        Returns:
+            object: The requested cached object.
+
+        Raises:
+            NotCached: When not found in the cache.
+        """
+        with self._lock:
+            try:
+                item = self._inventory[key]
+            except KeyError:
+                raise NotCached()
+            item.touch()
+            self.evict(self.eviction_threshold)
+            return item.object
+
+    def evict(self, threshold):
+        """
+        Evict all unused cached objects.
+
+        Args:
+            threshold (timedelta): Evict all cached objects that have not been
+                requested since the specified timedelta.
+
+        Returns:
+            list: The evicted objects.
+        """
+        evicted = []
+        now = Item.now()
+        with self._lock:
+            for key, item in self._inventory.items():
+                duration = (now - item.last_requested)
+                if item.busy or (duration < threshold):
+                    continue
+                self.purge(key)
+                evicted.append(item.object)
+        return evicted
+
+    def __iter__(self):
+        return self._inventory.iterkeys()
+
+
+class Item(object):
+    """
+    A cached item.
+    Contained within the cache inventory and is used to
+    track status and usage statistics.
+
+    Attributes:
+        last_requested (datetime): The last time the object was requested.
+        object (object): The actual cached object.
+    """
+
+    @staticmethod
+    def now():
+        """
+        The current timestamp.
+
+        Returns:
+            datetime: The current timestamp.
+        """
+        return datetime.utcnow()
+
+    def __init__(self, object):
+        """
+        Args:
+            object (object): The actual cached object.
```

Comment:
Instead of introducing an odd semantic, I think I'll just change to `object_` as recommended by PEP8 for name collisions w/ builtin _globals_.

https://www.python.org/dev/peps/pep-0008/#id36